### ISSUE TITLE
Inconsistent behaviour of SSA in combination with resourceVersion for different resources

### ISSUE THREAD

**Issue by felixhuettner** [kubernetes/kubernetes#106745]
### What happened?

I want to create new resources (for this example configmaps and services) using ServerSideApply. When creating these resources i want to make sure that the name of the resource is not already in use (as the name is generated randomly).

To create a new configmaps i follow the following workflow:
1. I generate a name that i wish to use
2. I list all configmaps in the namespace i wish to use
3. If the list already contains a resource with the name i want to have i start again from the beginning
4. I update my resource with the new data
   1. I set `.metadata.name` to the name i wish to use
   2. I set `.metadata.resourceVersion` to the `.metadata.resourceVersion` returned by the list request
5. I submit the request as PATCH with Content-Type `application/apply-patch+yaml`

If nobody took the name between the list and the PATCH call then the resource will be created an my fieldmanager will take `Apply` Ownership of the filed. If somebody took the name in the mean time then the PATCH request will fail with a HTTP 419.

<details>
  <summary>Example Script</summary>

Note that this skips step 3 above.

```bash
curl -X DELETE "http://localhost:8080/api/v1/namespaces/default/configmaps/tmp"

VERSION=$(curl "http://localhost:8080/api/v1/namespaces/default/configmaps" 2>/dev/null | jq -r ".metadata.resourceVersion")
echo $VERSION

curl -X PATCH "http://localhost:8080/api/v1/namespaces/default/configmaps/tmp?fieldManager=test123" -H "Content-Type: application/apply-patch+yaml" -H "Accept: application/json" -d @- <<EOF
{
    "kind":"ConfigMap",
    "apiVersion":"v1",
    "metadata":{
        "name":"tmp",
        "namespace":"default",
        "resourceVersion": "$VERSION"
    },
    "data":{
        "foo":"bar"
    }
}
EOF
```

</details>

This behaviour is the same for:
* Secrets
* PVCs
* Jobs
* Deployments
* PodDisruptionBudgets
* CustomResources

However if i do the same thing with a Service it fails with the error message `resourceVersion should not be set on objects to be created`.

<details>
  <summary>Example Script</summary>

Note that this skips step 3 above.

```bash
curl -X DELETE "http://localhost:8080/api/v1/namespaces/default/services/tmp"

VERSION=$(curl "http://localhost:8080/api/v1/namespaces/default/services" 2>/dev/null | jq -r ".metadata.resourceVersion")
echo $VERSION

curl -X PATCH "http://localhost:8080/api/v1/namespaces/default/services/tmp?fieldManager=test123" -H "Content-Type: application/apply-patch+yaml" -H "Accept: application/json" -d @- <<EOF
{
    "kind":"Service",
    "apiVersion":"v1",
    "metadata":{
        "name":"tmp",
        "namespace":"default",
        "resourceVersion": "$VERSION"
    },
    "spec":{
        "ports": [{
            "port": 8080,
            "targetPort": 8080
        }]
    }
}
EOF
```

</details>

Outputs:
```json
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "resourceVersion should not be set on objects to be created",
  "code": 500
}
```

### What did you expect to happen?

I expected that the behaviour is consistent across all resources.
As i am not sure if this is actually intended behaviour i can not say if the creation of resources with a `resourceVersion` should always work or always fail.

### How can we reproduce it (as minimally and precisely as possible)?

Run `kubectl proxy --port=8080` and use the two scripts from the problem description

### Anything else we need to know?

I am actually unsure if my inteded behaviour is covered under the guarantees set in https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#concurrency-control-and-consistency as the correlation between `resourceVersion` on objects and lists is not discussed there.

### Kubernetes version

<details>

```console
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.4", GitCommit:"3cce4a82b44f032d0cd1a1790e6d2f5a55d20aae", GitTreeState:"clean", BuildDate:"2021-08-11T18:10:22Z", GoVersion:"go1.16.7", Compiler:"gc", Platform:"linux/amd64"}
```

Tested also with `1.22.4` on k3s. Behaviour was the same.

</details>

### Cloud provider

<details>
Locally installed using kubeadm or k3s
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="CentOS Linux"
VERSION="8"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="CentOS Linux 8"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:8"
HOME_URL="https://centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"
CENTOS_MANTISBT_PROJECT="CentOS-8"
CENTOS_MANTISBT_PROJECT_VERSION="8"
$ uname -a
Linux xxx.xxx.xxx4.18.0-305.25.1.el8_4.x86_64 #1 SMP Wed Nov 3 10:29:07 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>
kubeadm, k3s
</details>

**Comment by liggitt** [comment]
is it only Service that displays the inconsistent behavior? is this reproducible in 1.22 (Service REST handling was unified with other resources in that version)

**Comment by felixhuettner** [comment]
It seems like it is only services. At least i did not find another resource that has this behaviour as well. Allthough i only tested the ones listed above.

Yes, i reproduced it using k3s in 1.22.4

**Comment by aojea** [comment]
/cc @aojea @thockin @khenidak 
If is only services we should take a look
/sig network

**Comment by felixhuettner** [comment]
I discovered the issue is no longer present on the master branch.
I will do a git bisect to find the fixing commit

**Comment by felixhuettner** [comment]
The issue is fixed in ccf337657075cd93619c25071f38840ce1acd3b8 .
So the fix will be included in 1.23.

I would assume this is a change that is not reasonably backportable?

**Comment by liggitt** [comment]
> The issue is fixed in [ccf3376](https://github.com/kubernetes/kubernetes/commit/ccf337657075cd93619c25071f38840ce1acd3b8) .
> So the fix will be included in 1.23.

oh, I didn't realize the service update change didn't land until 1.23. I would not expect ccf3376 to be backported, but a more targeted bugfix for this issue against earlier versions might be possible.

**Comment by felixhuettner** [comment]
We found a suitable workaround for older versions. So from my perspective we would not need a backport anymore.
Would it then make sense to close this issue?

---

For anybody having the same issue and interested in our solution:
1. Create the service using a POST call. Set a fieldmanager in this call.
2. The object is now created in k8s and has managed fields set to the `Update` operation
3. Send a SSA PATCH call that updates the newly created service. You must use the exactly same fields+values as in step 1.
4. The object now has two managed field entries with the same content. One for `Update` and one for `Apply`
5. Send a JSON-PATCH to update the service. It should look something like the following (replace $index with the index of the `Update` operation in `managedFields`
```
[{"op": "remove",
  "path": "/metadata/managedFields/$index",
}]
```
6. the object should now only have the `Apply` operation. The `Update` operation should be fully gone
